runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -679,6 +685,12 @@ protected final void invokeFinishBundle() {
       pushbackDoFnRunner.finishBundle();
       elementCount = 0L;
       lastFinishBundleTime = getProcessingTimeService().getCurrentProcessingTime();
+      // callback only after current bundle was fully finalized
+      // it could start a new bundle, for example resulting from timer processing
+      if (bundleFinishedCallback != null) {
+        bundleFinishedCallback.run();
+        bundleFinishedCallback = null;
```

Comment:
Could you share more from your findings? It looks like the only difference is that `elementCount/lastFinishBundleTime` is reset before calling `bundleFinishedCallback`. I see that starting a new bundle can cause problems because it may be finalized immediately due to the element count not being reset yet.

Reply:
On our 2.11 branch the callback fails because the current bundle isn't finished and a new one won't be started (stacktrace below). Even in master it will fail (though probably even harder to reproduce for the reason you mention).

I would like the callback to be in `DoFnOperator` because it is otherwise hard to understand and easy to re-introduce issues. As side effect, it can also be used elsewhere for other purposes. 

```
TimerException{java.lang.RuntimeException: Failed to process pushed back watermark after finished bundle.}
    at org.apache.flink.streaming.runtime.tasks.SystemProcessingTimeService$RepeatedTriggerTask.run(SystemProcessingTimeService.java:335)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Failed to process pushed back watermark after finished bundle.
    at org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator.lambda$processWatermark$0(ExecutableStageDoFnOperator.java:488)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator$SdkHarnessDoFnRunner.finishBundle(ExecutableStageDoFnOperator.java:632)
    at org.apache.beam.runners.flink.metrics.DoFnRunnerWithMetricsUpdate.finishBundle(DoFnRunnerWithMetricsUpdate.java:87)
    at org.apache.beam.runners.core.SimplePushbackSideInputDoFnRunner.finishBundle(SimplePushbackSideInputDoFnRunner.java:118)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.invokeFinishBundle(DoFnOperator.java:679)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.checkInvokeFinishBundleByTime(DoFnOperator.java:673)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.lambda$open$1(DoFnOperator.java:378)
    at org.apache.flink.streaming.runtime.tasks.SystemProcessingTimeService$RepeatedTriggerTask.run(SystemProcessingTimeService.java:330)
    ... 7 more
Caused by: java.lang.NullPointerException
    at org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator$SdkHarnessDoFnRunner.onTimer(ExecutableStageDoFnOperator.java:597)
    at org.apache.beam.runners.flink.metrics.DoFnRunnerWithMetricsUpdate.onTimer(DoFnRunnerWithMetricsUpdate.java:77)
    at org.apache.beam.runners.core.SimplePushbackSideInputDoFnRunner.onTimer(SimplePushbackSideInputDoFnRunner.java:112)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.fireTimer(DoFnOperator.java:720)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator.fireTimer(ExecutableStageDoFnOperator.java:394)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.onEventTime(DoFnOperator.java:703)
    at org.apache.flink.streaming.api.operators.HeapInternalTimerService.advanceWatermark(HeapInternalTimerService.java:288)
    at org.apache.flink.streaming.api.operators.InternalTimeServiceManager.advanceWatermark(InternalTimeServiceManager.java:108)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.processWatermark1(DoFnOperator.java:572)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.DoFnOperator.processWatermark(DoFnOperator.java:542)
    at org.apache.beam.runners.flink.translation.wrappers.streaming.ExecutableStageDoFnOperator.lambda$processWatermark$0(ExecutableStageDoFnOperator.java:486)
    ... 14 more
```

Reply:
The stack trace makes perfect sense. The bundle is finalized (via the flag), then started due to the watermark processing, then finalized immediately due to the element count not being reset. The exception is thrown because there is no active bundle. For the future, it would be helpful to share details directly in the JIRA.

Due to the complexity of DoFnOperator I agree it is nicer to process the callback there, instead of the subclass. I think we could add a test in a follow-up to harden the semantics of the callback.